### PR TITLE
Key value starter code

### PR DESCRIPTION
**Pull request by wsuppiger**
I am having trouble with Cmake, so hopefully we can go over cmake testing in class

### CODE REVIEW

**Comment by wsuppiger** [comment]
Also I just realized I made a silly mistake of KeyValue::void instead of KeyValue::functionName in keyValue.cc, but that is fixed now.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  GIT_TAG        release-1.10.0
+)
+
+
+# add_executable(app
+#   src/keyValue.cc)
+#
+# target_compile_features(app PRIVATE cxx_std_17)
+# target_compile_options(app PRIVATE -Wall -Wextra -Wpedantic)
+
```
Comment (barath) [inline CMakeLists.txt:21]: Remove blank lines.

File: src/keyValue.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2021, USC
```
Comment (barath) [inline src/keyValue.h]: Generally avoid camel case in file names.

File: src/keyValue.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// All rights reserved.
+#ifndef SRC_KEYVALUE_H_
+#define SRC_KEYVALUE_H_
+#include <unordered_map>
+#include <vector>
+
+namespace backend {
+
+// TODO: CHANGE BYTES TYPE
+typedef char BYTES;
```
Comment (barath) [inline src/keyValue.h]: What is this for?

File: src/keyValue.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2021, USC
+// All rights reserved.
+#ifndef SRC_KEYVALUE_H_
+#define SRC_KEYVALUE_H_
```
Comment (barath) [inline src/keyValue.h]: Add a blank line after the define.

File: src/keyValue.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+#include <unordered_map>
+#include <vector>
+
+namespace backend {
+
+// TODO: CHANGE BYTES TYPE
+typedef char BYTES;
+
+class KeyValue {
```
Comment (barath) [inline src/keyValue.h]: Add a class comment.

File: src/keyValue.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ public:
+  KeyValue() {}
```
Comment (barath) [inline src/keyValue.h]: Add an explicit initializer list for member vars.

File: src/keyValue.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+  ~KeyValue() {}
```
Comment (barath) [inline src/keyValue.h]: Make destructors virtual.

File: src/keyValue.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+  void put(const BYTES& key, const BYTES& value);
```
Comment (barath) [inline src/keyValue.h]: Capitalize method names.

Also, are these the right argument types?

File: src/keyValue.h [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  void get(const BYTES& key) const;
+  void remove(const BYTES& remove);
+
+ private:
+  std::unordered_map<BYTES, std::vector<BYTES>> storage_;
```
Comment (barath) [inline src/keyValue.h]: Add a member comment to describe the var.

File: src/keyValue_test.cc [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+#include "keyValue.h"
+
+#include <gtest/gtest.h>
+
+namespace backend {
+// test initialization of backend
+TEST(CreateKeyValue, Initialization) {
+  KeyValue backend;
+  backend.put('a', 'b');
```
Comment (barath) [inline src/keyValue_test.cc]: This doesn't appear to actually test what was put into the store. Add an EXPECT or ASSERT.

File: src/keyValue.h (hunk 2 above)
Comment (wsuppiger) [reply src/keyValue.h]: fixing in next commit

File: src/keyValue.h (hunk 3 above)
Comment (wsuppiger) [reply src/keyValue.h]: I was confused about how bytes worked in grpc, but @adamegyed explained to me today that it compiles to a stl string in C++, so I will change char to std::string.

File: src/keyValue.h (hunk 4 above)
Comment (wsuppiger) [reply src/keyValue.h]: fixing in next commit

File: src/keyValue_test.cc (hunk 10 above)
Comment (wsuppiger) [reply src/keyValue_test.cc]: Yeah this was a bad test.  I will add more sensible testing in the next PR

File: src/keyValue.h (hunk 8 above)
Comment (wsuppiger) [reply src/keyValue.h]: should I not use a typedef and just switch to string?

Comment (barath) [reply src/keyValue.h]: You should either use string or use a template; it's better not to have a typedef type unless you really need it for convenience.

Comment (wsuppiger) [reply src/keyValue.h]: ok makes sense

**Pull request merged**